### ISSUE TITLE
Ability to change endpoints from the BeforeExecution Event

### ISSUE THREAD

**Issue by MohitKumar1991** [schemathesis/schemathesis#514]
Hello again,
We have been using schemathesis for our production live API testing. One of the things we found was missing in the library was that we wanted to create a Resource and then later delete it at the end. For that we simply used the BeforeExecution Event and changed the endpoint parameters with the id of the Resource that was created.

Unfortunately in the master branch, this seems to have been removed, possibly to support subprocess running. I understand that this might be better for ThreadPoolRunner and SubProcessRunner but this closes off powerful use cases like ours.

My proposal is to move the ThreadPoolRunner to an AIOHttpRunner, so requests can still be made in parallel but the endpoints can be modified by user code before actual execution.

I am happy to help in any way possible as now our production API testing depends on this library.

**Comment by Stranger6667** [comment]
Hi @MohitKumar1991 !
Sorry for not getting back to you sooner. The main reason was to decouple events from the internals, so they can be used only to transfer the information from the runner to the consumer. It makes possible a subprocess runner (that is WIP now) so every event can be serialized and then used by e.g. frontend. Events were not intended to be used like that but I think that it is an interesting use case and I'd be happy to provide a way to support this behavior in the current version of the runner.

> My proposal is to move the ThreadPoolRunner to an AIOHttpRunner, so requests can still be made in parallel but the endpoints can be modified by user code before actual execution.

Unfortunately, this won't help since the initial goal of decoupling and making events serializable won't be fulfilled. In other words, this issue more on the outside part of the runner than on the inside one. Internally it is ok for threads to share schema or app instances, and they can work concurrently with non-serializable events. But if we want to share these events with different processes then we need to choose serialization format. It is quite annoying to play with pickling and restore their states after unpickling, so I went with a simpler approach of having JSON-serializable data types inside of them. And it would be nice to have the same API for both in-process and sub-process running modes, thus we need something more generic.

There is a PR (#520 ) that adds a stateful testing feature, which might be helpful for having an id of a created object in the next request to e.g. change it. I didn't look at it yet, but we'll likely have this feature in the future

Another approach could be having some hooks with an internal state that will allow modifying `Case` instances before they are executed. So, we can get an existing response and then update the generated case with some parameters. In this case, it will be helpful to have some way to manipulate the order in which tests are executed, so we can arrange a chain of create -> get -> update -> delete actions and make the abovementioned hook possible to work in all cases.

Hooks are also requested here - #458 which is kinda similar to what you need.

Another idea is to have an input queue, to control the runner, I didn't think much about it, but it might work.

Thank you for sharing this, I appreciate it. There is some exploration & experimenting work lies ahead and I'd be happy to share the results. I'll adjust priorities and will spend some time working on this feature next week. And when I'll share some prototypes / more detailed thoughts I'd be happy to have your feedback on it :)

Cheers

**Comment by MohitKumar1991** [comment]
Thank you so much for your reply and for assigning a high priority.

The hooks idea looks good enough for our first use case of changing the path_parameters.

For multiprocessing, is it possible to just serialize the options to the runner and the hooks functions and run everything in a separate process ? That way we won't need to serialize everything. The TestResultSet object can be serialized back to the main process. This way later this can be used with something like Celery.

Also is it possible to add the response in the AfterExecution Event ? and some summary of response in the TestResult Object ? The ones that we are most interested in are response code and response times. This way we will be able to feed TestResultSet into ExtentReport or ReportPortal.

Also will you accept any PRs on the master branch ? We have actually forked the repo and are running a forked version of Schemathesis Version 1.0.5 on our end and we would love to contribute back.

**Comment by MohitKumar1991** [comment]
Also just wanted to point out `https://pypi.org/project/multiprocess/` project which uses dill instead of pickle which is completely pythonic way of serializing. We use this in production.

**Comment by Stranger6667** [comment]
> For multiprocessing, is it possible to just serialize the options to the runner and the hooks functions and run everything in a separate process ?

Yes, the previous implementation was passing `app` instance to `prepare`, but apps can contain arbitrary objects inside that could be not-serializable in general. So it was changed to accept an importable path to the app. Also, there are hooks, they work when passed to a multiprocessing queue. 
The input is OK-ish for now, but the output contained the same references to app / schema which are not serializable. Even if they are it doesn't matter since the changes will be applied to a new deserialized object. In this sense even if we have everything safely serializable and deserializable we'll still have the problem of controlling another process.
Also, I tend to move away from Python-specific serialization primitives and use simpler objects which also solves a possible security issue with pickle or similar protocols. Another point is that in the future I'd like to have an alternative CLI implementation (in Rust) and these Python implementation details will be irrelevant there. So, the goal is to make the runner more independent and decoupled from everything else. Thank you for the link it is definitely an interesting tool that I'll try (at least it seems to solve some issues I was facing in my initial take on subprocess running mode)
The whole subprocess thing was done to avoid having a CPU-bound load in the web-worker process which will be awful performance-wise since it will take away resources from the web-worker, threading can help a bit because of the thread switching which will give some time to the web worker (CPU load is inside pure-python code), but it is not that scalable in the end.

> That way we won't need to serialize everything

I removed the events attributes that were not used in the CLI, now everything there is used

In general, communication, based on language-independent messaging seems more robust and scalable to me, but we are not yet there and there are quite a lot of challenges ahead of us.

> Also is it possible to add the response in the AfterExecution Event ?

It should be there now, we added `AfterExecution.result.interactions`, this information is not gathered by default but controlled with `store_interactions` boolean flag to the `runner.prepare` function. It is based on the VCR format and there are some thoughts regarding it in this PR (#462). This version is not released yet, but I hope to have it around Monday.
Let me know if there is anything we can add as extra to this format (`elapsed` response time is recorded there as well)

> Also will you accept any PRs on the master branch ?

Let's discuss it! I'd be happy to have contributions and feedback :) The most comfortable way I see here is that each separate feature is posted in a separate issue (to have it self-contained) so we agree on applicability of the idea, how it can be generalized for all use cases, which parts should be in the Schemathesis core and what is on the external side, etc. Then a PR, and implementation discussions there :)

Let me know if you need more information about the points above :)

**Comment by Stranger6667** [comment]
Closing it, as some hooks were implemented as a replacement, + #700 is somehow related. Otherwise, I consider the original discussion done.
